### PR TITLE
Kv group auth

### PR DESCRIPTION
**Pull request by suprjinx**

### CODE REVIEW

File: doc/openapi/paths/secrets.yml [hunk 1]
```diff
@@ -15,6 +15,10 @@ postSecrets:
                 type: string
                 description: "Path where the secret is stored"
                 example: "secret/storage/path"
+              gropus:
```
Comment (GeorgeJahad) [inline doc/openapi/paths/secrets.yml:18]: groups?

Comment (suprjinx) [reply doc/openapi/paths/secrets.yml]: fixed

File: app/lib/clients/vault/oidc.rb [hunk 2]
```diff
@@ -19,6 +19,24 @@ def get_oidc_client_config
         client.logical.read("auth/oidc/config")
       end
 
+      def create_oidc_role(role_name, groups, policy_name)
+        client.logical.write("auth/oidc/role/#{role_name}",
```
[the remark this answers is not in the document]
Comment (suprjinx) [reply app/lib/clients/vault/oidc.rb:23]: I was reading somewhere that this doesn't work for "external groups" defined in another identity provider (oidc)

[the remark this answers is not in the document]
Comment (suprjinx) [reply app/lib/clients/vault/oidc.rb:23]: but I will investigate -- maybe a group alias can make the link

File: README.md [hunk 3]
```diff
@@ -162,9 +162,6 @@ config/astral.yml).
 
 The rails test's configure the OIDC initial user, so if the tests pass,
 you can invoke the oidc login as follows:
-
-To use SSL in production, provide the necessary environment (SSL_CERT, SSL_KEY) to
-the container environment, and use the `bin/ssl.sh` startup command. Eg:
```
Comment (GeorgeJahad) [inline README.md:167]: was this comment supposed to go somewhere else?

File: app/lib/clients/vault/oidc.rb [hunk 4]
```diff
@@ -19,6 +19,24 @@ def get_oidc_client_config
         client.logical.read("auth/oidc/config")
       end
 
+      def create_oidc_role(role_name, read_groups, policy_name)
```
Comment (suprjinx) [inline app/lib/clients/vault/oidc.rb:22]: i will just remove role methods -- unused now

File: app/lib/utils/oidc_provider.rb [hunk 5]
```diff
@@ -81,9 +81,9 @@ def create_userpass_for_initial_user
   end
 
   def map_userpass_to_entity
-    entity = vault_client.logical.read(
+    group = vault_client.logical.read(
```
Comment (suprjinx) [inline app/lib/utils/oidc_provider.rb:84]: oops, rename caught too much... fixing

Comment (GeorgeJahad) [inline app/lib/utils/oidc_provider.rb:84]: this seems odd.  why the name change?

File: test/lib/clients/vault/policy_test.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  test "#verify_policy permits identity having the policy" do
+    policy_name = "some/policy/name"
+    @client.expects(:get_entity_data).with(@identity.sub).returns([ [ policy_name ], nil ])
+    assert_nil @client.verify_policy(@identity, policy_name)
+  end
+
+  test "#verify_policy looks for a role corresponding to consumer policy when supplied" do
+    producer_policy = "some/policy/name"
+    consumer_policy = "some/policy/other"
+    @client.expects(:get_entity_data).with(@identity.sub).returns([ [], nil ])
```
Comment (suprjinx) [inline test/lib/clients/vault/policy_test.rb:27]: adjusting so description is what happens here...

File: README.md (hunk 3 above)
Comment (suprjinx) [reply README.md:167]: it was duplicated

File: app/lib/clients/vault/policy.rb [hunk 7]
```diff
@@ -38,8 +53,27 @@ def remove_policy(identity, policy_name)
...
+
+      def remove_groups_policy(groups, policy_name)
+        groups.each do |group|
+          remove_group_policy(group, policy_name)
+        end
+      end
+
       private
 
+      def make_role_name(policy_name)
```
Comment (GeorgeJahad) [inline app/lib/clients/vault/policy.rb:73]: still needed?

File: app/lib/clients/vault/policy.rb [hunk 8]
```diff
@@ -2,14 +2,15 @@ module Clients
...
+      extend Oidc
 
       def rotate_token
         create_astral_policy
         token = create_astral_token
         Clients::Vault.token = token
       end
 
-      def assign_policy(identity, policy_name)
+      def assign_identity_policy(identity, policy_name)
```
Comment (GeorgeJahad) [inline app/lib/clients/vault/policy.rb:13]: I think a better name would be "assign_entity_policy"

File: app/models/secret.rb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+class Secret < ApplicationRecord
```
Comment (GeorgeJahad) [inline app/models/secret.rb:1]: How about KvMetadata instead of Secret?

File: app/lib/clients/vault/key_value.rb [hunk 10]
```diff
@@ -4,20 +4,35 @@ module KeyValue
...
+        # only producer can replace existing secret
+        if client.kv(kv_mount).read(path)
+          verify_policy(identity, producer_policy_path(path))
+        end
+
+        create_kv_policies(path)
+        assign_identity_policy(identity, producer_policy_path(path))
+        assign_groups_policy(read_groups, consumer_policy_path(path))
+        s = client.logical.write("#{kv_mount}/data/#{path}", data: data)
+        Secret.find_or_create_by(path: path).update(owner: identity.sub, read_groups: read_groups)
```
Comment (GeorgeJahad) [inline app/lib/clients/vault/key_value.rb:22]: Would it be better to write the metadata first?  (i.e. line 22 before line 21)?

File: app/lib/clients/vault/key_value.rb [hunk 11]
```diff
@@ -36,21 +51,34 @@ def kv_engine_type
...
-        client.sys.put_policy(policy_path(path), kv_policy(path))
+      def create_kv_policies(path)
+        client.sys.put_policy(producer_policy_path(path), kv_producer_policy(path))
+        client.sys.put_policy(consumer_policy_path(path), kv_consumer_policy(path))
       end
 
-      def policy_path(path)
-        "kv_policy/#{path}"
+      def producer_policy_path(path)
+        "kv_policy/#{path}/producer"
```
Comment (GeorgeJahad) [inline app/lib/clients/vault/key_value.rb:60]: "kv_policy/" seems a bit generic.  Do we need to call it "astral_kv_policy" to be safe?

File: app/lib/clients/vault/entity.rb [hunk 12]
```diff
@@ -8,6 +8,49 @@ def put_entity(name, policies, metadata = {})
...
 
+      def put_group(name, policies, auth_method = "oidc", metadata = {})
+        Domain.with_advisory_lock(name) do
+          original_policies, original_metadata = get_group_data(name)
+          policies = (policies || []) + original_policies
+          metadata = metadata.merge((original_metadata || {}))
+          client.logical.write("identity/group",
+                               {
+                                 name: name,
+                                 type: "external",
```
Comment (GeorgeJahad) [inline app/lib/clients/vault/entity.rb:19]: we shouldn't assume this is "external".  I think it is ok for now to throw an exception if it is not external, but I think it is dangerous to just overwrite it without checking.

File: app/lib/clients/vault/entity.rb [hunk 13]
```diff
@@ -8,6 +8,49 @@ def put_entity(name, policies, metadata = {})
...
+                                 policies: policies.uniq,
+                                 metadata: metadata
```
Comment (GeorgeJahad) [inline app/lib/clients/vault/entity.rb:21]: There are also the "member_entity_ids" and "member_group_ids" fields.  I saw an external group in B++2++ today that had those fields populated.  I'm not sure if those are generated or not.  Probably deserves some investigation.

File: app/lib/clients/vault/entity.rb [hunk 14]
```diff
@@ -8,6 +8,49 @@ def put_entity(name, policies, metadata = {})
...
+                               }
+                              )
+
+          # create group_alias if needed
+          unless original_policies.present?
```
[the remark this answers is not in the document]
Comment (GeorgeJahad) [reply app/lib/clients/vault/entity.rb:26]: that sounds good to me, something like this:
https://github.com/G-Research/astral/blob/main/app/lib/clients/vault/entity_alias.rb#L42-L44

File: app/lib/clients/vault/entity.rb [hunk 15]
```diff
@@ -19,11 +62,16 @@ def delete_entity(name)
...
+          [ [], {} ]
         else
           [ entity.data[:policies], entity.data[:metadata] ]
         end
       end
+
+      def read_group_alias(entity_name, alias_name)
+        id = read_entity_alias_id(entity_name, alias_name)
+        client.logical.read("identity/entity-alias/id/#{id}")
+      end
```
Comment (GeorgeJahad) [inline app/lib/clients/vault/entity.rb:74]: I think all uses of "entity" above should be "group"

File: app/lib/clients/vault/entity.rb (hunk 12 above)
Comment (suprjinx) [reply app/lib/clients/vault/entity.rb:19]: okay will ponder this -- maybe best to read and patch only policies if previously extant

File: app/lib/clients/vault/key_value.rb (hunk 10 above)
Comment (suprjinx) [reply app/lib/clients/vault/key_value.rb:22]: if creating secret in Vault fails do we want the metadata?

File: app/lib/clients/vault/policy.rb (hunk 7 above)
Comment (suprjinx) [reply app/lib/clients/vault/policy.rb:73]: removed

File: app/lib/utils/oidc_provider.rb [hunk 16]
```diff
@@ -81,9 +81,9 @@ def create_userpass_for_initial_user
   end
 
   def map_userpass_to_entity
-    entity = vault_client.logical.read(
+    identity = vault_client.logical.read(
```
Comment (GeorgeJahad) [inline app/lib/utils/oidc_provider.rb:84]: is this change needed?

File: app/lib/clients/vault/identity.rb [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+        client.logical.delete("identity/entity/name/#{name}")
+      end
+
+      private
+
+      def write_identity(path:, name:, policies:, defaults: {}, extra_params: [], merge_policies: true)
+        full_path = "#{path}/name/#{name}"
+        Domain.with_advisory_lock(full_path) do
+          identity = read_identity(full_path)
+          policies = (policies || []) + (identity&.data&.fetch(:policies) || []) if merge_policies
```
Comment (GeorgeJahad) [inline app/lib/clients/vault/identity.rb:41]: what is the advantage of using "fetch(:policies)" over "[:policies]" here?

File: app/lib/clients/vault/identity_alias.rb [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+      def put_entity_alias(entity_name, alias_name, auth_method)
+        write_identity_alias("entity", entity_name, alias_name, auth_method)
+      end
+
+      def put_group_alias(group_name, auth_method)
+        write_identity_alias("group", group_name, group_name, auth_method)
+      end
+
+      def read_group_alias(group_name, alias_name)
+        id = read_entity_alias_id(group_name, alias_name)
```
Comment (GeorgeJahad) [inline app/lib/clients/vault/identity_alias.rb:13]: read_group_alias_id??

File: app/lib/clients/vault/identity_alias.rb [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+        client.logical.read("identity/entity-alias/id/#{id}")
```
Comment (GeorgeJahad) [inline app/lib/clients/vault/identity_alias.rb:14]: group-alias?

File: app/lib/clients/vault/identity.rb (hunk 17 above)
Comment (suprjinx) [reply app/lib/clients/vault/identity.rb:41]: allows the `&.` (if not nil) object navigation on a hash

File: app/lib/clients/vault/identity_alias.rb (hunk 18 above)
Comment (suprjinx) [reply app/lib/clients/vault/identity_alias.rb]: i think this diff isn't right? don't see in the source

File: app/lib/clients/vault/identity_alias.rb (hunk 19 above)
Comment (suprjinx) [reply app/lib/clients/vault/identity_alias.rb]: i think this diff isn't right? don't see in the source

File: app/lib/clients/vault/key_value.rb (hunk 11 above)
Comment (suprjinx) [reply app/lib/clients/vault/key_value.rb]: uses the `kv_mount` (kv_astral) var to form this now.

File: app/lib/clients/vault/entity.rb (hunk 12 above)
Comment (suprjinx) [reply app/lib/clients/vault/entity.rb]: now external by default but takes the extant value where it exists

File: app/lib/clients/vault/entity.rb (hunk 13 above)
Comment (suprjinx) [reply app/lib/clients/vault/entity.rb]: added these as `extra_params` for writing the entity/group

File: app/lib/clients/vault/entity.rb (hunk 15 above)
Comment (suprjinx) [reply app/lib/clients/vault/entity.rb]: fixed

File: app/lib/utils/oidc_provider.rb (hunk 16 above)
Comment (suprjinx) [reply app/lib/utils/oidc_provider.rb:84]: reverted

File: app/lib/clients/vault/policy.rb [hunk 20]
```diff
@@ -1,43 +1,81 @@
...
+          put_entity(sub, [ policy_name ])
           put_entity_alias(sub, email, "oidc")
         end
       end
 
-      def verify_policy(identity, policy_name)
+      def assign_groups_policy(groups, policy_name)
+        groups.each do |group|
+          put_group(group, [ policy_name ])
+          put_group_alias(group, "#{group}-alias", "oidc")
```
Comment (GeorgeJahad) [inline app/lib/clients/vault/policy.rb:25]: I'm pretty sure that the group_alias has to be the group name for this to work with the oidc provider.

File: app/lib/clients/vault/identity.rb [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+
+      def put_group(name, policies)
+        write_identity(path: "identity/group",
+                       name: name,
+                       policies: policies,
+                       extra_params: [ :metadata, :type, :member_group_ids, :member_entity_ids ],
+                       defaults: { type: "external" })
+      end
+
+      def read_entity(sub)
```
Comment (GeorgeJahad) [inline app/lib/clients/vault/identity.rb:19]: since we are using the "entity/name" endpoint, I would change the param name to "name"

File: app/lib/clients/vault/identity.rb [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+        client.logical.read("identity/entity/name/#{sub}")
+      end
+
+      def delete_entity(name)
+        client.logical.delete("identity/entity/name/#{name}")
+      end
+
+      def get_entity_data(sub)
```
Comment (GeorgeJahad) [inline app/lib/clients/vault/identity.rb:27]: same as above

File: app/lib/clients/vault/identity_alias.rb [hunk 23]
```diff
@@ -0,0 +1,80 @@
+module Clients
+  class Vault
+    module IdentityAlias
+      def put_entity_alias(entity_name, alias_name, auth_method)
```
Comment (GeorgeJahad) [inline app/lib/clients/vault/identity_alias.rb:4]: In this file we use "auth_method" and "auth_path" interchangeably.  I think we should just use one.  auth_path is actually more accurate, I think.

File: test/models/domain_test.rb [hunk 24]
```diff
@@ -25,14 +25,4 @@ def setup
     assert_not @domain.valid?
     assert_includes @domain.errors[:fqdn], "can't be blank"
   end
-
```
Comment (GeorgeJahad) [inline test/models/domain_test.rb:28]: why did these tests get removed?

File: test/lib/clients/vault/identity_alias_test.rb [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+  end
+
+  test "#put_group_alias skips an existing group_alias" do
+    existing_alias = SecureRandom.hex
+    assert_raise {  @client.read_group_alias(@group_name, existing_alias, @auth_path) }
+    @client.put_group(@group_name, @policies)
+    assert_kind_of Vault::Secret, @client.put_group_alias(@group_name, existing_alias, @auth_path)
+    group_alias = @client.read_group_alias(@group_name, existing_alias, @auth_path)
+    assert_not_nil group_alias
+
```
Comment (GeorgeJahad) [inline test/lib/clients/vault/identity_alias_test.rb:65]: i would reread the group and confirm the alias is there, at this point.

File: app/lib/clients/vault/key_value.rb (hunk 10 above)
Comment (GeorgeJahad) [reply app/lib/clients/vault/key_value.rb]: I was thinking it is better to have the metadata without the secret, than the reverse.

File: test/models/domain_test.rb (hunk 24 above)
Comment (suprjinx) [reply test/models/domain_test.rb:28]: the `groups_array` function was removed (Domain.groups is now serialized as array automatically)

File: test/lib/clients/vault/identity_alias_test.rb (hunk 25 above)
Comment (suprjinx) [reply test/lib/clients/vault/identity_alias_test.rb:65]: ok will add

File: app/lib/clients/vault/identity.rb (hunk 21 above)
Comment (suprjinx) [reply app/lib/clients/vault/identity.rb:19]: done

File: app/lib/clients/vault/identity.rb (hunk 22 above)
Comment (suprjinx) [reply app/lib/clients/vault/identity.rb:27]: done

File: app/lib/clients/vault/identity_alias.rb (hunk 23 above)
Comment (suprjinx) [reply app/lib/clients/vault/identity_alias.rb:4]: done

File: test/lib/clients/vault/identity_alias_test.rb (hunk 25 above)
Comment (suprjinx) [reply test/lib/clients/vault/identity_alias_test.rb:65]: done

**Review by GeorgeJahad — APPROVED** [review]
I know I had a lot of comments on this one, but I think it really is better now.  Thanks!
